syntax = "proto3"; // 声明了protobuf的版本

package fixbug; // 声明了代码所在的命名空间

// 定义下面的选项，表示生成service服务器和rpc方法描述，默认不生成
option cc_generic_services = true;

message ResultCode
{
    int32 errcode = 1;
    bytes errmsg = 2; 
}

// 定义登录请求消息类型 name pwd
message LoginRequest
{
    bytes name = 1;
    bytes pwd = 2;
    // map<int32, string> test = 3; 
}

// 定义登录响应消息类型
message LoginResponse
{
    ResultCode result = 1;
    bool success = 2;
}

// 定义获取好友列表请求消息类型
message GetFriendListRequest
{
    int32 userid = 1;
}

// 定义用户消息类型
message User
{
    bytes name = 1;
    uint32 age = 2;
    enum Sex
    {
        MAN = 0;
        WOMAN = 1;
    }
    Sex sex = 3;
}

// 定义获取好友列表响应消息类型
message GetFriendListResponse
{
    ResultCode result = 1;
    repeated User friend_list = 2; // 使用repeated关键字 定义了一个列表类型
}

// 在protobuf里面怎么定义描述rpc方法的类型 - service
// 当你在protobuf里面想要定义描述rpc方法的服务类的时候 需要加一个可选项 否则不会生成cpp服务类
service UserServiceRpc
{
    rpc Login(LoginRequest) returns(LoginRequest);
    rpc GetFriendList(GetFriendListRequest) returns(GetFriendListResponse);
}